chain/action_test_helpers.go

```diff
@@ -0,0 +1,78 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package chain
+
+import (
+	"context"
+	"testing"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/stretchr/testify/require"
+
+	"github.com/ava-labs/hypersdk/codec"
+	"github.com/ava-labs/hypersdk/state"
+)
+
+type ActionTest struct {
+	Action Action
+
+	Rules     Rules
+	State     state.Mutable
+	Timestamp int64
+	Actor     codec.Address
+	ActionID  ids.ID
+
+	ExpectedOutputs   [][]byte
+	ExpectedErr       error
+	ContainsErrString string
+}
+
+type ActionTestSuite struct {
+	Tests map[string]ActionTest
+
+	Setup    func()
+	Teardown func()
+}
+
+func (suite *ActionTestSuite) Run(t *testing.T) {
+	if suite.Setup != nil {
+		suite.Setup()
+	}
+
+	for testName := range suite.Tests {
+		t.Run(testName, func(t *testing.T) {
+			require := require.New(t)
+			test := suite.Tests[testName]
+
+			// Ensure that only one of outputs or err is set
+			if test.ExpectedOutputs != nil && (test.ExpectedErr != nil || len(test.ContainsErrString) > 0) {
+				t.Fatalf("ExpectedOutputs and ExpectedErr are set in test %s", testName)
+			}
+			if test.ExpectedOutputs == nil && test.ExpectedErr == nil && len(test.ContainsErrString) == 0 {
+				t.Fatalf("neither ExpectedOutputs nor ExpectedErr is set in test %s", testName)
+			}
+
+			output, err := test.Action.Execute(context.TODO(), test.Rules, test.State, test.Timestamp, test.Actor, test.ActionID)
+
+			if err != nil {
+				if test.ExpectedErr == nil && len(test.ContainsErrString) == 0 {
+					t.Fatalf("got unexpected error: %w", err)
+				}
+
+				if test.ExpectedErr != nil {
+					require.Equal(err, test.ExpectedErr)
+				}
+				if len(test.ContainsErrString) > 0 {
+					require.Contains(err.Error(), test.ContainsErrString)
+				}
+			}
+
+			require.Equal(output, test.ExpectedOutputs)
```

Comment:
err and output should be mutually exclusive but it doesn't hurt to still check it at the end, even if `err != nil`

Reply:
You won't reach the output check because you're using `require`; swapping to `assert` will use `t.Errorf()` under the hood instead of `t.Fatalf()`, but I don't think you should.

> it doesn't hurt to still check it at the end, even if `err != nil`

I think it does actually, even if very minor, because it forces every error path to have a strict return value (presumably `nil`) when in reality the value is undefined.

[More context](https://google.github.io/styleguide/go/decisions#keep-going) should anyone be interested.

Reply:
> I think we should keep using require, otherwise the test author needs to keep thinking about which failures will cause the test to fail vs where the test can continue.

That's actually the point. It avoids whack-a-mole fixes where you get partial information about problems. The test author really should be thinking about that because they're the one person with the greatest insight.

My rule of thumb is that errors use `require` because that almost certainly guarantees that all other returned values are incorrect. Incorrect `actual` or `got` values, however, are typically more suited to `assert`.

> Additionally, require is more concise than if err != nil style checks, which is needed to use t.Fatal.

`require` is equivalent to `t.Fatal` so there's no need for the `err != nil` check. The alternative is `assert`, which is equivalent to `t.Error`.